internal/services/oracledatabase/exadata_infrastructure_resource.go

```diff
@@ -0,0 +1,374 @@
+				}
+			}
+
+			if err := client.CreateOrUpdateThenPoll(ctx, id, param); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r ExadataInfraResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+
+			client := metadata.Client.OracleDatabase.OracleDatabaseClient.CloudExadataInfrastructures
+			id, err := cloudexadatainfrastructures.ParseCloudExadataInfrastructureID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var model ExadataInfraResourceModel
+			if err = metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding err: %+v", err)
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving exists when updating: +%v", *id)
+			}
+			if existing.Model == nil && existing.Model.Properties == nil {
+				return fmt.Errorf("retrieving as nil when updating for %v", *id)
+			}
+
+			if metadata.ResourceData.HasChangesExcept("tags") {
+				return fmt.Errorf("only `tags` currently support updates")
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				update := &cloudexadatainfrastructures.CloudExadataInfrastructureUpdate{
+					Tags: tags.Expand(model.Tags),
+				}
+				err = client.UpdateThenPoll(ctx, *id, *update)
+				if err != nil {
+					return fmt.Errorf("updating %s: %v", id, err)
+				}
+			}
+			return nil
+		},
+	}
+}
+
+func (ExadataInfraResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			id, err := cloudexadatainfrastructures.ParseCloudExadataInfrastructureID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			client := metadata.Client.OracleDatabase.OracleDatabaseClient.CloudExadataInfrastructures
+			result, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(result.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return err
+			}
+
+			if result.Model == nil {
+				return fmt.Errorf("retrieving %s got nil model", id)
+			}
+
+			prop := result.Model.Properties
+			output := ExadataInfraResourceModel{
+				CustomerContacts:  FlattenCustomerContacts(result.Model.Properties.CustomerContacts),
+				Name:              pointer.ToString(result.Model.Name),
+				Location:          result.Model.Location,
+				Zones:             result.Model.Zones,
+				ResourceGroupName: id.ResourceGroupName,
+				Tags:              utils.FlattenPtrMapStringString(result.Model.Tags),
+				ComputeCount:      pointer.From(prop.ComputeCount),
+				DisplayName:       prop.DisplayName,
+				StorageCount:      pointer.From(prop.StorageCount),
+				Shape:             prop.Shape,
+				MaintenanceWindow: FlattenMaintenanceWindow(prop.MaintenanceWindow),
+			}
```

Comment:
We don't error in the read when the `Model` is nil, similar to the data source we just nil check it to prevent a crash. Also I think in the data source we nil checked `Properties` which means it's a pointer and should be nil checked here as well
```suggestion		
			state := ExadataInfraResourceModel{
				Name: id.Name
				ResourceGroupName: id.ResourceGroupName
			}
			
			if model := result.Model; model != nil {
				state.Location = location.Normalize(model.Location)
				state.Zones = model.Zones
				state.Tags = model.Tags
				if props := model.Properties; props != nil {
					state.CustomerContacts = FlattenCustomerContacts(props.CustomerContacts)
					... etc.
				}
			}

```

Reply:
Done, thanks.